File: pkg/earthsci-ast-ts/src/const-units.test.ts
/**
 * Declared units on a `const` node (esm-spec §4.8.5): a `const` that declares its
 * units has that unit, a `const` without units stays undeterminable, and the
 * field is gated at esm 1.2.0.
 */
import { describe, it, expect } from 'vitest'
import { parseUnitForConversion } from './unit-conversion.js'
import type { ParsedUnit } from './unit-conversion.js'
import { checkDimensions } from './units.js'
import { rejectConstUnitsPreV12 } from './solver.js'
import type { Expression } from './types.js'

const bindings = (units: Record<string, string>): Map<string, ParsedUnit> =>
  new Map(Object.entries(units).map(([name, u]) => [name, parseUnitForConversion(u)]))

const konst = (units?: string): Expression =>
  ({
    op: 'const',
    args: [],
    value: 0.44704,
    ...(units === undefined ? {} : { units }),
  }) as unknown as Expression

describe('declared units on a const node', () => {
  it('gives a const with declared units that unit', () => {
    const env = bindings({ speed_mph: 'mi/h' })
    const got = checkDimensions(
      { op: '*', args: ['speed_mph', konst('m*h/(mi*s)')] } as Expression,
      env,
    )
    const ms = parseUnitForConversion('m/s')
    expect(got.dimensions).not.toBeNull()
    expect(got.dimensions?.exact.equals(ms.exact)).toBe(true)
  })

  it('leaves a const without units undeterminable', () => {
    const env = bindings({ speed_mph: 'mi/h' })
    const got = checkDimensions({ op: '*', args: ['speed_mph', konst()] } as Expression, env)
    expect(got.dimensions).toBeNull()
  })

  it('rejects declared units in a document below esm 1.2.0, naming the node', () => {
    const doc = (esm: string) => ({
      esm,
      models: {
        M: { equations: [{ lhs: 'x', rhs: { op: 'const', args: [], value: 1, units: 'm' } }] },
      },
    })
    expect(() => rejectConstUnitsPreV12(doc('1.1.0'))).toThrow(/models\/M\/equations\/0\/rhs/)
    expect(() => rejectConstUnitsPreV12(doc('1.2.0'))).not.toThrow()
  })
})

File: pkg/EarthSciAST.jl/test/const_units_test.jl
# Declared units on a `const` node (esm-spec §4.8.5): a `const` that declares its
# units has that unit, a `const` without units stays undeterminable, an
# unresolvable declared unit is listed for the structural layer, and the field is
# gated at esm 1.2.0.

using Test
using EarthSciAST
using Unitful

include("testutils.jl")

@testset "declared units on a const node (esm-spec §4.8.5)" begin
    konst(units) = EarthSciAST.OpExpr("const", EarthSciAST.ASTExpr[]; value = 0.44704, units = units)
    units = Dict("speed_mph" => "mi/h")

    @testset "a const with declared units has that unit" begin
        got = EarthSciAST.get_expression_dimensions(_op("*", _v("speed_mph"), konst("m*h/(mi*s)")), units)
        @test got !== nothing
        @test EarthSciAST._exact_scale(got) == EarthSciAST._exact_scale(EarthSciAST.parse_units("m/s"))
    end

    @testset "a const without units stays undeterminable" begin
        @test EarthSciAST.get_expression_dimensions(_op("*", _v("speed_mph"), konst(nothing)), units) === nothing
    end

    @testset "unresolvable const units are listed" begin
        @test EarthSciAST.unresolvable_const_units(_op("*", _v("speed_mph"), konst("mph"))) == ["mph"]
        @test isempty(EarthSciAST.unresolvable_const_units(_op("*", _v("speed_mph"), konst("m*h/(mi*s)"))))
    end

    @testset "const units are gated at esm 1.2.0" begin
        doc(esm) = Dict{String,Any}("esm" => esm, "models" => Dict{String,Any}("M" => Dict{String,Any}(
            "equations" => Any[Dict{String,Any}("lhs" => "x", "rhs" => Dict{String,Any}(
                "op" => "const", "args" => Any[], "value" => 1.0, "units" => "m"))])))
        err = try
            EarthSciAST.reject_const_units_pre_v12(doc("1.1.0"))
            nothing
        catch e
            e
        end
        @test err !== nothing
        @test occursin("/models/M/equations/0/rhs", sprint(showerror, err))
        @test EarthSciAST.reject_const_units_pre_v12(doc("1.2.0")) === nothing
    end
end
